### DISCUSSION TITLE
Why are module names not properly highlighted?

### DISCUSSION THREAD

**Discussion by bzm3r** [haskell/haskell-language-server#4055] [Q&A] [answered]
I am using the VS Code extension to integrate HLS.

I import a module called `Bird1`, and in the picture below, we can see textmate scopes being picked up but no semantic token type:

![image](https://github.com/haskell/haskell-language-server/assets/4350451/ed62eee3-ce34-420d-81c4-6db26749cc7b)

Similarly, when I hover over the usage of `Bird1` in the file, it gets picked up as a *variable* instead of a module:
![image](https://github.com/haskell/haskell-language-server/assets/4350451/bcd7bf4c-78d3-4ce2-a474-4609ea872b12)

Note that a name following a `module` token is also not recognized by the semantic highlighter:

![image](https://github.com/haskell/haskell-language-server/assets/4350451/c26590b4-0ebe-45ec-a86c-4cc4bcafade6)

It seems to me that the requisite functionality does exist:
https://github.com/haskell/haskell-language-server/blob/0047d133a11049958c78f076da45f0dfcaf492bb/plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Types.hs#L43

Am I right in understand that? So, there's something missing from the VS Code Extension's end? (I think if I am right in understanding the issue, then there I know what to do to fix it in the VS Code extension.)

(I am also happy to contribute fixes to HLS, if necessary! Some guidance on where I might want to start looking would be helpful.)

**Comment by soulomoon** [comment, marked as answer]
Which version are you using, module name splitting out is only available in the master head.

we currently have

<img width="647" alt="image" src="https://github.com/haskell/haskell-language-server/assets/14073857/eea21a9d-a794-4847-9c62-583436e28ccd">

**Reply by bzm3r** [reply]
Ah, I am not using the master version! Okay, I think that explains the issue.

**Comment by michaelpj** [comment]
@soulomoon did you also do highlighting of the module name in `module X where`?

**Reply by soulomoon** [reply]
No, it seems does not appear in the hieAst where we get our identifiers

**Reply by michaelpj** [reply]
Maybe we should make a ticket for that. It's not very important, but it will make things look a little inconsistent I guess.

**Reply by soulomoon** [reply]
agree

**Reply by soulomoon** [reply]
I also observed some inconsistent with `applicativedo`, since it modifies the ast during the transition `parsedSource -> typecheckSource`. 
I am actually considering about using the parsedSource for names and modules and collect whatever we can in occname first.
Then combine this with the hieAst to create semantic tokens. (similar proccess is happening in the rust analyzor if i am not mistaken, but they have their lossless syntax trees [rowan](https://github.com/rust-analyzer/rowan)).

This way we have some advantages.
1. Even if the file fail to typecheck for some reason, we might still be able to have some highlighting.
2. Avoid some breaks introduced by extension like `applicativedo`.

**Comment by soulomoon** [comment]
@michaelpj  I have created an issue for the missing module name in `module x where` in ghc https://gitlab.haskell.org/ghc/ghc/-/issues/24493

**Comment by soulomoon** [comment]
Also proposed a PR in ghc to fix that https://gitlab.haskell.org/ghc/ghc/-/merge_requests/12153
